### DISCUSSION TITLE
Proposal: Quasi-quotation

### DISCUSSION THREAD

**Discussion by declard** [dotnet/csharplang#3104] [Language Ideas]
Currently C# is very limited in code quotation abilities as it only supports lambda-expressions without statements in them which were originally introduced for LINQ.
Some other languages like LISP and Nemerle provide a much more powerful tool - quasi-quotation.
I.e. not only they allow to represent language expressions (syntax trees) statically as literals, but also allow to incorporate dynamically generated expressions into them in the same manner as string interpolation works.
That's how it could possibly look in C#:
```csharp
Expression F(Func<int, Expression> expr)
{
  var fourtyTwo = Expression.Contant(42);
  return ${ x => (x + $fourtyTwo) * $(expr(1)) };
}
```
In this case `F(i => Expression.Constant(i))` will return an expression equivalent to `x => (x + 42) * 1`

In the code above `${ expression }` represents a quote with a literal representation of an expression inside, `$name` and `$(expression)` represent incorporated values

Potentially a quotation block could contain any expression representable by the `System.Linq.Expressions` hierarchy.
That means quotes can be recursive - quotes within quotes, which questions the rules of interpolation: should `$name` always target the topmost (in quotation sense) name and be escaped with subsequent appending of more `$`s, or should it "unquote" one level up and each `$` added means "unqoute one more level up":
```csharp
var e = expr1;
var lam = ${ e => ${ $e } };
lam.Compile().Invoke(expr2);
```
In the first case the result would be `expr1`, in the second case `expr2`

**Comment by orthoxerox** [comment]
This feature should probably depend on [Roslyn syntax tree API](https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.syntaxtree?view=roslyn-dotnet) instead of `System.Linq.Expressions`, as the latter is feature frozen and will never cover everything.

Have you seen [this](https://github.com/dotnet/roslyn/blob/features/source-generators/docs/features/source-generators.md) document?

**Reply by declard** [reply]
Is it still frozen considering https://github.com/dotnet/csharplang/discussions/4727?

**Comment by declard** [comment]
> This feature should probably depend on Roslyn syntax tree API

Not necessarily. It depends on which subset of the language is supposed to be supported by the quasi-quotation. It could potentially work with `System.Linq.Expressions`, even though with some restrictions. Or it could even work with both `System.Linq.Expressions` and `Microsoft.CodeAnalysis.SyntaxTree` (like lambda expressions can be treated as either `Func<>`s or `Expression<>`s).

> Have you see this document?

I haven't seen that document, but it looks to be about source code generation, which is the opposite to quasi-quotation (expression to text vs text to expression)

**Comment by orthoxerox** [comment]
As I've said, `System.Linq.Expressions` already doesn't cover any expressions added to the language in the last few releases and it never will. Associating your proposal with this API confers practically no benefits.

As for your comparison with source generation, I don't consider them diametrically opposite. Quasiquotation is not really text to expression, it's AST to expression, I expect the compiler to transform the quote into a bunch of calls to the syntax factory methods at compile time. This is something that source generators will benefit from as well, as the MVP approach of passing a big string to the compiler means practically zero generator-side checks. Passing an AST will ensure that at least the syntax is correct. So to me both quasi-quotation and source generation are about turning ASTs to code, except one does that at run time and the other at compile time.

**Comment by YairHalberstadt** [comment]
The quasi quotation API in Scala is one of the most confusing and frustrating ones to use.

I don't know if that's because Scala has a badly designed api, or because quasi quotation in languages that aren't designed for it is difficult. Either way I feel that a lot of design work would need to go into this to make sure it is pleasant to use, or at least not distinctly unpleasant.

**Comment by orthoxerox** [comment]
To save everyone a moment: https://docs.scala-lang.org/overviews/quasiquotes/intro.html

**Comment by orthoxerox** [comment]
Also, I wonder if Razor syntax can be adapted to work with qq's. Your current syntax doesn't let you include control flow statements inside quotes, while Razor has things like `@foreach` and `@if` that let you splice different nodes into the result.

**Comment by svick** [comment]
You don't need new syntax for this, what you need is [LINQKit](https://github.com/scottksmith95/LINQKit). Using that, your example code would be:

```c#
Expression<Func<int, int>> F(Func<int, Expression<Func<int>>> expr)
{
    var fourtyTwo = Expression.Lambda<Func<int>>(Expression.Constant(42));
    var expr1 = expr(1);
    return Linq.Expr((int x) => (x + fourtyTwo.Invoke()) * expr1.Invoke()).Expand();
}

…

// returns "x => ((x + 42) * 1)"
F(i => Expression.Lambda<Func<int>>(Expression.Constant(i))).ToString()
```

While this is significantly more code, it works today and doesn't require changing the language for a niche use case.

**Comment by declard** [comment]
@svick the essense of the proposal is to avoid operating expressions explicitly and to write code "as-is" instead, but LINQKit seems to do some part of a job in terms of Linq.Expression hierarchy

**Comment by svick** [comment]
@declard But my example is not really operating on expressions explicitly. Notice that my code uses e.g. `+` and not `Expression.Add`.

The main differences from your proposal are:

* To interpolate an expression, it has to be a `LambdaExpression`, not just an `Expression`.
* The syntax to interpolate is `expr.Invoke()` instead of `$expr`.
* The syntax for the whole interpolation is `Linq.Expr(…).Expand()` instead of `${ … }`.

**Comment by Blealtan** [comment]
@svick The `Linq.Expr` you mentioned is making use of Expression Trees, which has various certain problems writing general programs, e.g., no block statement is allowed in the inner lambda expression. Also, the Expression Tree should be now deprecated and will not update any further if I'm correct.

Scala 3.0 is introducing a [new metaprogramming scheme](https://dotty.epfl.ch/docs/reference/metaprogramming/toc.html), which employs a quote operator `${ ... }` and a splice operator `'{ ... }`, and keep the code phase-consistent (thus reasonable and correct) through the so-called PCP (Phase Consistency Principle). Combined with their new public syntax tree API (TASTy, which stands for Typed AST) and a corresponded set of staging API (making it possible to compile quotes at runtime), it makes it easy for libraries to do both compile-time and runtime code generation. I'm personally trying to build an eDSL with this new feature and find it really comfortable compared to previous languages with metaprogramming I've used, including Rust procedural macros, Scala 2.x macros and quasiquotes, C++ templates, and also C# Expression Trees.

For C#:
1. <del> Roslyn already provides a rich and stable set of SyntaxTree API (like Scala 3.0 TASTy); </del> Roslyn BoundTree is currently implementation detail as mentioned by @orthoxerox, so stabilize and make it public would be required;
1. Roslyn can already compile C# code at runtime for later execution (like Scala 3.0 `quoted.staging`);
1. `Linq.Expression` already exists as a fully-typed expression data structure (like Scala 3.0 `quoted.Expr`).

So we should at least be able to do what Scala 3.0 is doing. **Introduce the two operators**, **generate Expression Tree objects for quoted codes**, **enable full-featured syntax in quoted codes**, <del>**update and back the old `Linq.Expression` with the new Roslyn SyntaxTree API**</del>**provide Bound Trees as a new set of API in Roslyn and back `Linq.Expression` with it**, and we will do it.

With full-featured metaprogramming, we can retire both Expression Trees and Code Generators, making things easier for various libraries using the two features, like Linq2Sql and serialization libraries I suppose.

While I do like the Scala 3.0 metaprogramming pretty much, I'm not proposing to do exactly what they do; rather, I'm mentioning it as a practical example on how a modern managed language design metaprogramming features. Anyway it'll be nothing but simulating Lisp Macros in a modern language; what matters is that HOW we simulate it. The current (while old) Expression Tree feature is not doing it well.

@declard In the example you've provided, it seems you're using `${ ... }` in both quoting and unquoting scenarios; it might be confusing in some way, but maybe we can also use the PCP rule reversely to infer which phase every quote belongs to and determine it should quote or unquote. I would still say it's confusing since it's not obvious to coders whether an expression is being quoted or not when the nesting is deep.

**Reply by declard** [reply]
Not really, `${ ... }` is for quoting and `$..` (with any valid expression, and expressions, as we know, don't begin with `{` yet) is for unquoting just like `` ` `` and `,` in Scheme respectively.

Want to unquote two levels - `$$..` and so on.

**Comment by orthoxerox** [comment]
@Blealtan TASTy is more similar to the bound trees of Roslyn, which are an implementation detail right now. SyntaxTree API covers only untyped syntax trees.

**Comment by Blealtan** [comment]
> @Blealtan TASTy is more similar to the bound trees of Roslyn, which are an implementation detail right now. SyntaxTree API covers only untyped syntax trees.

Thanks for pointing that out! I've modified my comment. Kinda sleepy lol

**Comment by svick** [comment]
@Blealtan 

> The `Linq.Expr` you mentioned is making use of Expression Trees

It does, because I was suggesting it as an alternative to the original proposal, which also uses Expression Trees.

Personally, I think that for run time metaprogramming, removing the limitations of Expression Trees is a better choice (and more likely to happen) than introducing a brand new API.

> With full-featured metaprogramming, we can retire […] Expression Trees

I'm not sure what "retire" means exactly, but I don't think Expression Trees are going to be deprecated or removed from .Net anytime soon, because they're used quite a lot by existing code.

> With full-featured metaprogramming, we can retire […] Code Generators

How? One of the most important aspects of Code Generators is that they run at compile time. As far as I can tell, what you're proposing is metaprogramming that happens at run time and that alone would mean it's not a suitable replacement for Code Generators.

---

Also, if you want to discuss your proposal in more detail, I think it would make sense to create a new issue for it, with more specific explanation of how it would work (e.g. including an example code that uses what you're proposing).

**Comment by Blealtan** [comment]
@svick Sure. I do have answers in response to your concerns, and I'll fire a new issue on my proposal when I have time (really busy these days XD)
